packages/cli/generation/remote-generation/remote-workspace-runner/src/SourceUploader.ts

```diff
@@ -0,0 +1,141 @@
+import { AbsoluteFilePath, join, RelativeFilePath } from "@fern-api/fs-utils";
+import { ApiDefinitionSource, SourceConfig } from "@fern-api/ir-sdk";
+import { loggingExeca } from "@fern-api/logging-execa";
+import { InteractiveTaskContext } from "@fern-api/task-context";
+import { FernWorkspace, IdentifiableSource } from "@fern-api/workspace-loader";
+import { FernRegistry as FdrAPI } from "@fern-fern/fdr-cjs-sdk";
+import { readFile, unlink } from "fs/promises";
+import tmp from "tmp-promise";
+
+const PROTOBUF_ZIP_FILENAME = "proto.zip";
+
+export type SourceType = "asyncapi" | "openapi" | "protobuf";
+
+export class SourceUploader {
+    private context: InteractiveTaskContext;
+    private sources: Record<string, IdentifiableSource>;
+    private sourceTypes: Set<SourceType>;
+
+    constructor(context: InteractiveTaskContext, workspace: FernWorkspace) {
+        this.context = context;
+        this.sources = Object.fromEntries(workspace.getSources().map((source) => [source.id, source]));
+        this.sourceTypes = new Set<SourceType>(Object.values(this.sources).map((source) => source.type));
+    }
+
+    public getSourceTypes(): Set<SourceType> {
+        return this.sourceTypes;
+    }
+
+    public getFdrApiDefinitionSources():
+        | Record<FdrAPI.api.v1.register.SourceId, FdrAPI.api.v1.register.Source>
+        | undefined {
+        if (Object.keys(this.sources).length === 0) {
+            return undefined;
+        }
+        return Object.fromEntries(
+            Object.entries(this.sources).map(([id, source]) => [
+                id,
+                {
+                    type: source.type === "protobuf" ? "proto" : source.type
+                }
+            ])
+        );
```

Comment:
can we annotate the return type as part of this function

Reply:
It's actually already annotated, it just ends up being formatted across multiple lines since it's a long type. An alternative would be to create a private alias for the type like so:

```typescript
type FdrSourceMap = Record<FdrAPI.api.v1.register.SourceId, FdrAPI.api.v1.register.Source>;

...

public getFdrApiDefinitionSources(): FdrSourceMap | undefined {
    if (Object.keys(this.sources).length === 0) {
        return undefined;
    }
    return Object.fromEntries(
        Object.entries(this.sources).map(([id, source]) => [
            id,
            {
                type: source.type === "protobuf" ? "proto" : source.type
            }
        ])
    );
}
```

Not sure it's worth it though. I like that this function handles the `undefined` case seamlessly within the function, but I'll see if I can refactor things with a `!= null` guard to prevent such a large signature in the first place.

Reply:
I was able to get rid of this entirely, so feel free to ignore 👍

Reply:
my b on this